src/polynomial.rs

```diff
@@ -1,343 +1,471 @@
-use crate::{fft_precompute, fft_with_precomputation, ifft_with_precomputation_power_of_2, util::log2_ceil, Field};
+use crate::{fft_precompute, fft_with_precomputation, ifft_with_precomputation_power_of_2, util::log2_ceil, AffinePoint, Curve, FftPrecomputation, Field, MsmPrecomputation, PolynomialCommitment};
+use std::cmp::Ordering;
+use std::ops::{Index, IndexMut, RangeBounds};
+use std::slice::{Iter, IterMut, SliceIndex};
 
-// Store polynomial as a list of coefficient starting with the constant coefficient.
-type Polynomial<F> = Vec<F>;
+/// Polynomial struct holding a polynomial in coefficient form.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Polynomial<F: Field>(Vec<F>);
 
-fn degree<F: Field>(a: &Polynomial<F>) -> usize {
-    let mut ans = 0;
-    a.iter().enumerate().for_each(|(i, x)| {
-        if x.is_nonzero() {
-            ans = i;
-        }
-    });
-    ans
+impl<F: Field> From<Vec<F>> for Polynomial<F> {
+    /// Takes a vector of coefficients and returns the corresponding polynomial.
+    fn from(coeffs: Vec<F>) -> Self {
+        Self(coeffs)
+    }
 }
 
-fn lead<F: Field>(a: &Polynomial<F>) -> F {
-    let mut ans = F::ZERO;
-    a.iter().for_each(|&x| {
-        if x.is_nonzero() {
-            ans = x;
-        }
-    });
-    ans
-}
+impl<F, I> Index<I> for Polynomial<F>
+where
+    F: Field,
+    I: SliceIndex<[F]>,
+{
+    type Output = I::Output;
 
-pub fn polynomial_is_zero<F: Field>(a: &Polynomial<F>) -> bool {
-    a.iter().all(|&x| x.is_zero())
+    /// Indexing on the coefficients.
+    fn index(&self, index: I) -> &Self::Output {
+        &self.0[index]
+    }
 }
 
-fn rev<F: Field>(a: &Polynomial<F>) -> Polynomial<F> {
-    assert!(a.last().unwrap().is_nonzero());
-    let mut r = a.to_vec();
-    r.reverse();
-    r
+impl<F, I> IndexMut<I> for Polynomial<F>
+where
+    F: Field,
+    I: SliceIndex<[F]>,
+{
+    fn index_mut(&mut self, index: I) -> &mut <Self as Index<I>>::Output {
+        &mut self.0[index]
+    }
 }
 
-fn neg<F: Field>(a: &Polynomial<F>) -> Polynomial<F> {
-    a.iter().map(|&x| -x).collect()
-}
+impl<F: Field> Polynomial<F> {
+    /// Takes a slice of coefficients and returns the corresponding polynomial.
+    pub fn from_coeffs(coeffs: &[F]) -> Self {
+        Self(coeffs.to_vec())
+    }
 
-pub fn trim<F: Field>(a: &mut Polynomial<F>) {
-    let d = degree(a);
-    a.drain(d + 1..);
-}
+    /// Returns the coefficient vector.
+    pub fn coeffs(&self) -> Vec<F> {
```

Comment:
Maybe best to return a reference and let the caller determine if it needs to be cloned or not. I think it'd be equally idiomatic since Rust getters often return refs, from what I've seen.